toolkit/offchain/src/permissioned_candidates.rs

```diff
@@ -1,15 +1,195 @@
+//! This UTXO should have 1 token of the Permissioned Candidates Policy with an empty asset name.
+//! The datum encodes Permissioned Candidates using VersionedGenericDatum envelope with the Permissioned Candidates stored
+//! in the `datum` field of it. Field should contain list of list, where each inner list is a triple of byte strings
+//! `[sidechain_public_key, aura_public_key, grandpa_publicKey]`.
 
-use crate::csl::{get_builder_config, InputsBuilderExt, TransactionBuilderExt, TransactionContext};
+use crate::csl::{
+	convert_ex_units, get_builder_config, InputsBuilderExt, TransactionBuilderExt,
+	TransactionContext,
+};
 use crate::plutus_script::PlutusScript;
+use anyhow::anyhow;
 use cardano_serialization_lib::{
-	ExUnits, JsError, Transaction, TransactionBuilder, TxInputsBuilder,
+	ExUnits, JsError, PlutusData, Transaction, TransactionBuilder, TxInputsBuilder,
 };
+use ogmios_client::query_ledger_state::QueryLedgerState;
+use ogmios_client::query_network::QueryNetwork;
+use ogmios_client::transactions::{OgmiosEvaluateTransactionResponse, Transactions};
 use ogmios_client::types::OgmiosUtxo;
-use partner_chains_plutus_data::permissioned_candidates::permissioned_candidates_to_plutus_data;
-use sidechain_domain::PermissionedCandidateData;
+use partner_chains_plutus_data::permissioned_candidates::{
+	permissioned_candidates_to_plutus_data, PermissionedCandidateDatums,
+};
+use sidechain_domain::{McTxHash, PermissionedCandidateData, UtxoId};
+
+pub async fn upsert_permissioned_candidates<C: QueryLedgerState + QueryNetwork + Transactions>(
+	genesis_utxo: UtxoId,
+	candidates: &Vec<PermissionedCandidateData>,
+	payment_signing_key: [u8; 32],
+	ogmios_client: &C,
+) -> anyhow::Result<Option<McTxHash>> {
+	let ctx = TransactionContext::for_payment_key(payment_signing_key, ogmios_client).await?;
+	let (validator, policy) =
+		crate::scripts_data::permissioned_candidates_scripts(genesis_utxo, ctx.network)?;
+	let validator_address = validator.address_bech32(ctx.network)?;
+	let validator_utxos = ogmios_client.query_utxos(&[validator_address]).await?;
+	let mut candidates = candidates.clone();
+	candidates.sort();
+
+	match get_current_permissioned_candidates(validator_utxos)? {
+		Some((_, current_permissioned_candidates))
+			if current_permissioned_candidates == *candidates =>
+		{
+			log::info!("Current permissioned candidates are equal to the one to be set.");
+			Ok(None)
+		},
+		Some((current_utxo, _)) => {
+			log::info!(
+				"Current permissioned candidates are different to the one to be set. Updating."
+			);
+			Ok(Some(
+				update_permissioned_candidates(&validator, &policy, &candidates, &current_utxo, ctx, ogmios_client)
+					.await?,
+			))
+		},
+		None => {
+			log::info!("There are permissioned candidates. Inserting new ones.");
+			Ok(Some(insert_permissioned_candidates(&validator, &policy, &candidates, ctx, ogmios_client).await?))
+		},
+	}
+}
+
+fn get_current_permissioned_candidates(
```

Comment:
AFAICT this is nearly identical to `get_current_d_parameter` so maybe they could be shared and renamed eg `get_validator_datum`

Reply:
It is but my DRY threshold is 3.

I could make it generic but this would require:
* type parameter itself
* parameter for constructing log messages

Also it would require to change the result type here to PermissionedCandidatesDatums and move sorting outside, but I don't really think it is a problem.